repackaged as zero-dependency typescript library

CedarMist opened the pull request:
This now includes ct32 implementation of AES and the uint32 library
Everything is ESM & CommonJS compatibile, with ESM tree-shaking support.
With a minified build this resolves to 10kb of code Moved from npm to pnpm

CedarMist commented:
Oops, ran into one problem: `"default" is not exported by `

```
import deoxysii from '@oasisprotocol/deoxysii';
```

pro-wh reviewed (commented):
lotta new stuff here. haven't looked through the moved code yet

so we're vendoring in uint32 and bsaes? is it that those libraries didn't have good tree shaking?

CedarMist commented on package.json:
```diff
@@ -1,34 +1,55 @@
+	"license": "MIT",
+	"author": "Oasis Protocol Foundation <info@oasisprotocol.org>",
+	"devDependencies": {
+		"@biomejs/biome": "^1.7.0",
```
Yup, as an alternative to eslint+prettier I find biomejs is faster and 'just works'.

CedarMist commented on tsconfig.base.json:
```diff
@@ -0,0 +1,45 @@
+		"importHelpers": true, // This is only used for build validation. Since we do not have `tslib` installed, this will fail if we accidentally make use of anything that'd require injection of helpers.
+
+		// Language and environment
+		"moduleResolution": "NodeNext",
```
This is copied from the Viem project: https://github.com/wevm/viem

pro-wh replied:
o pretty substantial, should we put in their license too?

CedarMist replied:
> Configuration files are not typically copyrightable as they are used to configure parameters and settings rather than express creative content. They are usually considered functional and utilitarian in nature rather than original works of authorship.

pro-wh replied:
whose quote

pro-wh replied:
https://github.com/babel/babel/blob/main/packages/babel-preset-typescript/package.json#L10

CedarMist replied:
The quote is a summary from GPT which lines up with my experience, a `tsconfig.json` file doesn't really meet the threshold of originality required for copyright protection, in the same way that a shopping list or recipe doesn't.

pro-wh replied:
oh that's referring to copyright not protecting that abstract configuration. they don't own the combination of enabling `strict` and setting `moduleResolution` to nodenext. this is file copied, and copyright is meant to protect fixed representations like that

pro-wh replied:
I'd like to leave my side of this copyright discussion there, and I'm glad we shared our perspectives on this. I'd like to add some other opinions not related to copyright:

1. I want readers to know that Viem is to credit for this configuration regardless of licensing requires it or not
2. I want our developers to know what this configuration is based on, in case we want to watch for upstream changes that we like too

CedarMist replied:
Yup lets add a permalink to the original viem `tsconfig.json` and call it a day.

CedarMist replied:
Re: "NodeNext" vs "nodenext", the JSONSchema referenced in the documentation at https://www.typescriptlang.org/docs/handbook/tsconfig-json.html uses `NodeNext`:
 * https://github.com/SchemaStore/schemastore/blob/6dd10d986ab375d7df73dce10b1cc3a7c285a63c/src/schemas/json/tsconfig.json#L421

pro-wh replied:
:eyes: woa thanks microsoft

CedarMist commented on tsconfig.base.json:
```diff
@@ -0,0 +1,45 @@
+		// Language and environment
+		"moduleResolution": "NodeNext",
+		"module": "NodeNext",
+		"target": "ES2021", // Setting this to `ES2021` enables native support for `Node v16+`: https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping.
```
No, Node 18+ only - https://endoflife.date/nodejs

Security updates stopped for Node 16 last year, active support ended 1.5 years ago.

pro-wh replied:
https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping should we target node 18 then?

CedarMist commented on package.json:
```diff
@@ -1,34 +1,55 @@
+		"type": "git",
+		"url": "https://github.com/oasisprotocol/deoxysii-js.git"
+	},
+	"license": "MIT",
```
deoxysii.ts & aes.ts are still MIT licensed

The uint32 package doesn't really have a license (and is explicitly noted by the author as something you can do whatever tf you want with it), so that's been noted in an appropriate `SPDX-License-Identifier`. I believe it would be entirely kosher to strip the license header, add my own copyright and re-license it as MIT if I wanted.

pro-wh replied:
o ok

pro-wh replied:
> strip the license header

sounds like that's allowed

> add my own copyright

I don't think the license controls whether you can do that, and I think you can't do that

> re-license it as MIT

I think you have enough permission to redistribute it under MIT, which is more restrictive

CedarMist replied:
It would be fair to add a copyright notice to it given that the changes I made to it are, while trivial, about as much effort as writing it in the first place.

Either way I'm not fussed, please advise on what to do with `uint32.ts`

pro-wh replied:
if it was done by changing the uint32.js albeit extensively, that sounds like it's a derivative work. I've read that it's usually preferred to contribute to a project under its original license, which for the uint32.js library was a custom license. so overall, this deoxysii-js package would have a custom license, of what would be expressed in pseudo-SPDX as (MIT AND uint32.js's license)

CedarMist commented on scripts/tsconfig.json:
```diff
@@ -0,0 +1,7 @@
+	"compilerOptions": {
+		"composite": true,
+		"module": "nodenext",
+		"types": ["bun-types"]
```
Yup this will be removed as it's unnecessary.

CedarMist commented on scripts/rename-cjs.mjs:
```diff
@@ -0,0 +1,15 @@
+import { promises as fs } from "node:fs";
+import url from "node:url";
+
+process.chdir(url.fileURLToPath(`${import.meta.url}/../../dist/_cjs`));
```
```typescript
console.log(import.meta.url)
console.log(url.fileURLToPath(`${import.meta.url}`))
console.log(url.fileURLToPath(`${import.meta.url}/../../dist/_cjs`))
```

=

```
file:///src/deoxysii-js/scripts/rename-cjs.mjs
/src/deoxysii-js/scripts/rename-cjs.mjs
/src/deoxysii-js/dist/_cjs
```

pro-wh replied:
woa neat. thanks for the walkthrough

CedarMist commented on package.json:
```diff
@@ -1,34 +1,55 @@
+		"vitest": "^1.5.3"
+	},
+	"peerDependencies": {
+		"typescript": ">=5.0.4"
```
This will reuse a common typescript version from one of `devDependencies`.

I should've added it to `devDependencies` though so it is not a production dependency, I have made that change.

pro-wh replied:
but if it's a dev-only dependency, then we already have full control over the dependency, so does this actually add any constraints?

CedarMist replied:
If we look at `tsup` as an example of good usage of peer dev dependencies.
`tsup`  has `5.0.2` as a `devDependency`, but they also have `typescript >=4.5.0` as a `peerDependency` with the `optional=true` flag set for typescript in `peerDependenciesMeta`. I assume 4.5.0 is the bare minimum version of typescript they want to support.

When I install `tsup` as a dev dependency their package will use whichever version of typescript I have installed. If `tsup` didn't have it as a peer dependency it's probable that I would now have two versions of typescript in the same project.

Then if I install `@oasisprotocol/deoxysii` as a devDependency of another project, with the peer dependency setup correctly, this will not potentially bring in another version of typescript, meaning typescript should act more predictably.

Is it entirely necessary... no. Do we have the same problem with the C#, OCaml, Rust, Java, PHP, C++ or C ecosystems... no (they do have their own problems, but none seem to be as asinine as the JS/Typescript/NPM ecosystem). Does Deno have the same problem... no.

Lets look at an example:

```
examples/hardhat-boilerplate/frontend$ pnpm list --depth=10 | wc -l
414739
```

Now, when people consistently use peer dependencies with typescript this makes the mess just a little less, for example:

```
examples/hardhat-boilerplate/frontend$ pnpm list --depth=10 | grep ' typescript '
│ └── typescript 4.9.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ └── typescript 5.4.5 peer
│ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ └── typescript 5.4.5 peer
│ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ │ └── typescript 5.4.5 peer
│ │ │ │ └── typescript 5.4.5 peer
│ └── typescript 5.4.5 peer
│ │ ├── typescript 5.4.5 peer
│ ├── typescript 5.4.5 peer
├── typescript 5.4.5 peer
```

lukaw3d replied:
uuh, i don't know if you're using something experimental but in all established tools:
- installing @oasisprotocol/deoxysii as a devDependency does not mean that deoxysii's devDependencies will get installed
- if @oasisprotocol/deoxysii specified typescript >=4.5.0 as a dependency that would be enough to get deduplicated with existing typescript 5.4.5 in a project
- Deno should have this problem or it will have much larger problems

CedarMist replied:
Ah, sorry I've just reached my frustration point with javascript stuff.

pro-wh replied:
my understanding of peer dependencies is that they're for use at runtime. deoxysii-js doesn't use typescript at runtime. I believe we shouldn't have this entry.

tsup, when it runs, uses the typescript compiler, thus it lists typescript as a runtime dependency. it lists it in `peerDependencies` instead of `dependencies` to make sure npm doesn't end up installing a distinct version just for tsup when other tools are using a different version of typescript.

CedarMist commented on scripts/rename-cjs.mjs:
```diff
@@ -0,0 +1,15 @@
+	const cjs = await fs.readFile(filename, "utf-8");
+	await fs.writeFile(
+		filename.replace(/\.js$/, ".cjs"),
+		cjs.replaceAll(/require\("(.\/[\w_\-.]+)\.js"\)/g, 'require("$1.cjs")'),
```
This doesn't remove the `./`, it's included in capture group `$1`

pro-wh replied:
oh true

lukaw3d commented on package.json:
```diff
@@ -1,34 +1,55 @@
 {
```
The PR title and description don't say what the motivation for these changes is
(and "zero-dependency" misleads me to think that some dependency was removed, but it wasn't)

CedarMist commented on src/deoxysii.ts:
```diff
@@ -0,0 +1,701 @@
+
+import { uint32_xor2, uint32_xor3 } from "./uint32.js";
+
+namespace deoxysii {
```
Ah, that's a good point.

Do you have any ideas on how to support both the default export and individual named imports? I thought a namespace would solve that, but obviously it doesn't

CedarMist replied:
Ah, I found a workaround, just added an `index.ts` file with:
```typescript
import { AEAD, ErrKeySize, ErrNonceSize, ErrOpen, KeySize, NonceSize, TagSize } from "./deoxysii";
export { AEAD, ErrKeySize, ErrNonceSize, ErrOpen, KeySize, NonceSize, TagSize };
const deoxysii = {
    AEAD,
    ErrKeySize,
    ErrNonceSize,
    ErrOpen,
    KeySize,
    NonceSize,
    TagSize
}
export default deoxysii;
```

lukaw3d replied:
acceptable. Probably equivalent to adding this in deoxysii.ts
```ts
export default {
    AEAD,
    ErrKeySize,
    ErrNonceSize,
    ErrOpen,
    KeySize,
    NonceSize,
    TagSize
};
```

pro-wh commented on src/uint32.ts:
```diff
@@ -0,0 +1,267 @@
+// SPDX-License-Identifier: WTFPL
```
to be clear, this is a relicensing

uint32.js license (https://github.com/fxa/uint32.js?tab=readme-ov-file#license) is:

> Do, what You want. This library was designed to be a copy/paste template. It would be nice to hear from You, if You used this library or if You just copy/pasted some source. It would be nice, if you added a "contains code of Franz X Antesberger" or something like that or a ref to the github project to your license information or elsewhere.

WTFPL (https://spdx.org/licenses/WTFPL.html) is:

> DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
Version 2, December 2004
>
> Copyright (C) 2004 Sam Hocevar <sam@hocevar.net>
>
> Everyone is permitted to copy and distribute verbatim or modified copies of this license document, and changing it is allowed as long as the name is changed.
>
> DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
>
> 0. You just DO WHAT THE FUCK YOU WANT TO.

some sentiment in common though

lukaw3d commented on src/deoxysii.ts:
```diff
@@ -0,0 +1,701 @@
+			stkDeriveK(key, this.derivedKs);
+		}
+
+		/**
+		 * @param {Uint8Array} nonce
+		 * @param {Uint8Array | null} plaintext
+		 * @param {Uint8Array | null} associatedData
+		 */
```
These comments aren't needed anymore and could drift from types

CedarMist commented:
Ok @lukaw3d and @pro-wh 

I have incorporated your comments to the best that I'm able to, thank you for the feedback as I'd missed or not considered a few things that you pointed out.

I've also updated the README so it's pretty and includes an example.

CedarMist commented on biome.json:
```diff
@@ -0,0 +1,12 @@
+{
```
Biome doesn't use `.eslintrc`, it uses `biome.json`

lukaw3d commented on scripts/rename-cjs.mjs:
```diff
@@ -0,0 +1,15 @@
+/** Renames `.js` files in the /lib/cjs folder to `.cjs` to appease ESM clients. */
+
+import { promises as fs } from "node:fs";
+import url from "node:url";
+
+process.chdir(url.fileURLToPath(`${import.meta.url}/../../dist/_cjs`));
+for (const filename of await fs.readdir(".")) {
+	if (!/\.js$/.test(filename)) continue;
+	const cjs = await fs.readFile(filename, "utf-8");
+	await fs.writeFile(
+		filename.replace(/\.js$/, ".cjs"),
+		cjs.replaceAll(/require\("(.\/[\w\-.]+)\.js"\)/g, 'require("$1.cjs")'),
+	);
+	await fs.unlink(filename);
+}
```
I tried comparing your previous version + without namespace + using tsup (https://antfu.me/posts/publish-esm-and-cjs#tsup)

https://github.com/oasisprotocol/deoxysii-js/commit/32e7271652ac6e2194b22d186c7d413e7da16c26

might be useful some day

CedarMist replied:
tsup looks really good, and I really don't like the `rename-cjs` hack. I wish there was a simpler way that only needed a single set of files.

Unfortunately there seems to be no safe way to drop CommonJS support

CedarMist replied:
```
$ pnpm tsup src/index.ts  --format cjs,esm --dts --clean --minify --sourcemap
CLI Building entry: src/index.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v8.0.2
CLI Target: es2021
CLI Cleaning output folder
CJS Build start
ESM Build start
ESM dist/index.mjs     10.17 KB
ESM dist/index.mjs.map 68.06 KB
ESM ⚡️ Build success in 21ms
CJS dist/index.js     10.69 KB
CJS dist/index.js.map 68.32 KB
CJS ⚡️ Build success in 22ms
DTS Build start
DTS ⚡️ Build success in 723ms
DTS dist/index.d.ts  1.17 KB
DTS dist/index.d.mts 1.17 KB
```

`tsup` is great, so much less hassle.

lukaw3d commented on package.json:
```diff
@@ -1,34 +1,67 @@
+		"distclean": "pnpm run clean && rm -rf node_modules *.tgz",
+		"build": "pnpm tsup src/index.ts --format cjs,esm --dts --clean",
+		"prepublishOnly": "npm run build",
+		"debug-imports-cjs": "NODE_DEBUG=module,esm bash -c 'pnpm node --experimental-global-webcrypto examples/check.cjs 2> >(grep deoxysii)'",
```
I don't understand this. Is it just
`NODE_DEBUG=module,esm node --experimental-global-webcrypto examples/check.cjs | grep deoxysii`
minus the errors?

CedarMist replied:
It greps stderr for logs from `module` or `esm` to show only ones related to deoxysii.

Essentially this double checks that check.cjs loads the .js file via the normal module loader, and check.mjs loads the .mjs file via ESM

CedarMist replied:
```

> @oasisprotocol/deoxysii@0.0.6 debug-imports /home/runner/work/deoxysii-js/deoxysii-js
> pnpm run debug-imports-cjs && pnpm run debug-imports-esm


> @oasisprotocol/deoxysii@0.0.6 debug-imports-cjs /home/runner/work/deoxysii-js/deoxysii-js
> NODE_DEBUG=module,esm bash -c 'pnpm node --experimental-global-webcrypto examples/check.cjs 2> >(grep deoxysii)'

MODULE 2161: Module._load REQUEST /home/runner/work/deoxysii-js/deoxysii-js/.pnpmfile.cjs parent: /home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/dist/pnpm.cjs
MODULE 2161: looking for "/home/runner/work/deoxysii-js/deoxysii-js/.pnpmfile.cjs" in ["/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/dist/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules","/home/runner/node_modules","/home/node_modules","/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/.node_modules","/home/runner/.node_libraries","/opt/hostedtoolcache/node/18.20.2/x64/lib/node"]
MODULE 2179: looking for "/home/runner/work/deoxysii-js/deoxysii-js/examples/check.cjs" in ["/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/.node_modules","/home/runner/.node_libraries","/opt/hostedtoolcache/node/18.20.2/x64/lib/node"]
MODULE 2179: load "/home/runner/work/deoxysii-js/deoxysii-js/examples/check.cjs" for module "."
MODULE 2179: Module._load REQUEST @oasisprotocol/deoxysii parent: .
MODULE 2179: looking for "@oasisprotocol/deoxysii" in ["/home/runner/work/deoxysii-js/deoxysii-js/examples/node_modules","/home/runner/work/deoxysii-js/deoxysii-js/node_modules","/home/runner/work/deoxysii-js/node_modules","/home/runner/work/node_modules","/home/runner/node_modules","/home/node_modules","/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/nocjs [class AEAD]
Encrypted: Uint8Array(27) [
  233,  24, 247,  21,  97, 209, 159,
  253,   5, 151, 186,  13, 204,  42,
  227, 216, 204, 121,  32, 176, 232,
   24,  16, 216, 130, 102, 206
]
Decrypted: Hello World
de_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/.node_modules","/home/runner/.node_libraries","/opt/hostedtoolcache/node/18.20.2/x64/lib/node"]
MODULE 2179: load "/home/runner/work/deoxysii-js/deoxysii-js/dist/index.js" for module "/home/runner/work/deoxysii-js/deoxysii-js/dist/index.js"

> @oasisprotocol/deoxysii@0.0.6 debug-imports-esm /home/runner/work/deoxysii-js/deoxysii-js
> NODE_DEBUG=module,esm bash -c 'pnpm node --experimental-global-webcrypto examples/check.mjs 2> >(grep deoxysii)'

star [class AEAD]
default [class AEAD]
AEAD [class AEAD]
Encrypted: Uint8Array(27) [
  175, 189, 113,  76, 235, 191, 199,
   63, 217, 208,  56, 109,  32, 214,
  207, 137, 242,   9,   9, 141,   6,
  220, 222,  46, 100, 147, 174
]
Decrypted: Hello World
MODULE 2204: Module._load REQUEST /home/runner/work/deoxysii-js/deoxysii-js/.pnpmfile.cjs parent: /home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/dist/pnpm.cjs
MODULE 2204: looking for "/home/runner/work/deoxysii-js/deoxysii-js/.pnpmfile.cjs" in ["/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/dist/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules","/home/runner/node_modules","/home/node_modules","/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/bin/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules/pnpm/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/pnpm@8.15.8/node_modules","/home/runner/setup-pnpm/node_modules/.pnpm/node_modules","/home/runner/.node_modules","/home/runner/.node_libraries","/opt/hostedtoolcache/node/18.20.2/x64/lib/node"]
ESM 2222: Storing file:///home/runner/work/deoxysii-js/deoxysii-js/examples/check.mjs (implicit type) in ModuleLoadMap
ESM 2222: Translating StandardModule file:///home/runner/work/deoxysii-js/deoxysii-js/examples/check.mjs
ESM 2222: Storing file:///home/runner/work/deoxysii-js/deoxysii-js/dist/index.mjs (implicit type) in ModuleLoadMap
ESM 2222: Translating StandardModule file:///home/runner/work/deoxysii-js/deoxysii-js/dist/index.mjs
```

lukaw3d commented on src/uint32.ts:
```diff
@@ -0,0 +1,267 @@
+// SPDX-License-Identifier: WTFPL
```
Split. Diff from original files is now visible in 36fa9714f0c207095c6b19bf4619e9ff5d999690
